### ISSUE TITLE
Copy Ratio adjustments for purity/ploidy are incorrect

### ISSUE THREAD

**Issue by tedtoal** [lima1/PureCN#40]
In some places in PureCN there are adjustments of copy ratio to account for purity and ploidy, or vice-versa, to take an adjusted value and return it to an unadjusted "observed" ratio value.  Marcus Riester gave the following reference for the method used to do this:

https://www.nature.com/articles/ng.2760 (section Impurity-corrected GISTIC)

The equations in that section contain an algebra mistake.  I'm concerned that PureCN might incorporate that mistake in one or more places.  I have verified that it uses the correct equation in at least two places.

Details:

The above reference shows this derivation:

```
  R(x) = (aq(x)+2(1-a))/D
  D = aT + 2(1-a)
  q(x) = DR(x)/a - 2(1-a)/a
  R'(x) = q(x)/T = R(x)/a - 2(1-a)/aT
```

where:

```
  R(x) = raw (observed) coverage ratio (PureCN's seg.mean, gene.mean, etc.)
  R'(x) = adjusted coverage ratio (in tumor cells)
  q(x) = integer copy number in cancer cells
  D = average ploidy across all cells of tumor (of sample)
  a = sample purity
  T = tumor ploidy
```

However, in the last step where q(x) is substituted in q(x)/T, the algebra is wrong.  The correct algebra is:

```
R'(x) = q(x)/T = DR(x)/aT - 2(1-a)/aT = (aT + 2(1-a))R(x)/aT - 2(1-a)/aT
      = R(x) + 2(1-a)R(x)/aT - 2(1-a)/aT
      = [aTR(x) + 2(1-a)R(x) - 2(1-a)]/aT
```

As a test, say that purity=a=0.5, tumor ploidy=T=2, and raw coverage ratio=1.5. Then we expect the adjusted coverage ratio to be 2 (tumor segment is 2X amplify (4 copies) and this becomes raw ratio of 1.5 when purity is 1/2 [0.5*4 + 0.5*2] / 2 = 1.5):

```
Paper:     R'(x) = 1.5/0.5 - 2(0.5)/(0.5 * 2) = 3 - 2(0.5) = 2 (correct)
Corrected: R'(x) = [0.5*2*1.5 + 2(0.5)1.5 - 2(0.5)] / (0.5*2) = 1.5 + 1.5 - 1 = 2 (correct)
```

But now suppose that tumor ploidy=T=4, and we still have purity=a=0.5.  Say raw coverage ratio=1.0, meaning there is no tumor amplification and the number of copies at any locus is the same as the mean number of copies, in both the 2X normal and 4X tumor tissue.  We expect the adjusted coverage ratio to also be 1:

```
Paper:     R'(x) = 1/0.5 - 2(0.5)/(0.5 * 4) = 2 - 2(0.5)/2 = 2 - 1/2 = 1.5 (wrong)
Corrected: R'(x) = [0.5*4*1 + 2(0.5)1 - 2(0.5)] / (0.5 * 4) = [2 + 1 - 1] / 2 = 2 / 2 = 1 (correct)
```

Marcus Riester indicated PureCN uses the following:

```
    rds <- readRDS("Sampleid.rds")
    r <- rds$results[[1]]
    r$seg$seg.mean.adjusted <- r$seg$seg.mean/r$purity - 2*(1-r$purity)/(r$purity*r$ploidy)
```

The above equation matches the incorrect one in the paper, so PureCN must be wrong wherever the above appears in the PureCN code, although I haven't found where that is (but I didn't search exhaustively).

The PureCN function .calcExpectedRatio() does the inverse operation, computing R(x) from R'(x), and it is using the correct equation.

PureCN function runAbsoluteCN() has this line:

```
    opt.C <- (2^(seg$seg.mean + log.ratio.offset) *  total.ploidy)/p - ((2 * (1 - p))/p)
```

and since tumor copy number C = tumor coverage ratio * tumor ploidy, this equation at first appears to be the paper's (incorrect) R'(x) * ploidy.  However, "total.ploidy" is probably the ploidy in the sample as a whole, which corresponds to the paper's "D", and in that case, the above equation is correct.  The algebra error only creeps in in the last derivation, R'(x) = ...

In summary, my concerns are (1) the above equation for seg.mean.adjusted (I don't know where in the code that is); (2) Could there be other places in PureCN where the algebra bug crept in?

**Comment by lima1** [comment]
Thanks Tod, I'll test the formula from the paper very soon. I've never really used it. I think the only need for an adjusted log-ratio is for tools like GISTIC that expect log-ratios. PureCN gives you the purity AND ploidy adjusted copy numbers.

I'm fairly certain that the code in the package is correct. What's maybe confusing in the runAbsoluteCN code is that there is ploidy and total.ploidy. Total ploidy includes the normal contamination and is in your last example 3 (0.5*4+0.5*2).

In log-space, that means:
(2^0*3)/0.5 - ((2 * (1 - 0.5))/0.5)=4

Since we did not know the ploidy when adjusting for coverage, a copy number of 4 is expected to have the same coverage as a normal copy number 2.
 
A tumor copy number of 2 has an expected log2 ratio of about -0.59
log2(PureCN:::.calcExpectedRatio(2, 0.5,4))
[1] -0.5849625

(2^(-0.585)*3)/0.5 - ((2 * (1 - 0.5))/0.5)=1.999896

**Comment by tedtoal** [comment]
Marcus,

First, Ted, not Tod… :-)

Second, I just want to point out that the algebra in the Zack paper is wrong, so if PureCN is correct, it is because you didn’t use the equation in that paper.

The Zack paper does not mention ploidy vs. total.ploidy, but it does mention two types of ploidy, which I believe are the same thing.  It uses T (tau) to refer to the tumor tissue ploidy, which I think it what you are referring to simply as “ploidy”.  It uses D to refer to average ploidy across all cells in the cancer, and I believe this is probably what you refer to as “total.ploidy”.

You said you were using the following equation:
r$seg$seg.mean.adjusted <- r$seg$seg.mean/r$purity - 2*(1-r$purity)/(r$purity*r$ploidy)
It contains “ploidy” and not “total.ploidy”, and it uses the incorrect equation from Zack.  I didn’t see that equation in the PureCN code, but then, I’ve only looked at a small part of the PureCN code.

The value C that appears in many of the PureCN output files is a purity- and ploidy-adjusted number (you said), and it appeared to me that in at least one place where C was being calculated, the seemingly incorrect equation was being used:

opt.C <- (2^(seg$seg.mean + log.ratio.offset) *  total.ploidy)/p - ((2 * (1 - p))/p)
However, this equation DOES contain total.ploidy.  If total.ploidy is equivalent to the Zack paper’s D, then the above equation can be rewritten using the Zack paper’s notation (with a=alpha/purity, T=tau=ploidy in tumor tissue) as:

opt.C <- (R(x) * D) / a - 2*(1-a)/a

The equation for C using the Zack equation in the paper is:

C = T * R’(x) = q(x) = integer copy number in cancer cells = D * R(x) / a - 2*(1-a)/a

Which is the same equation as PureCN’s.  That equation is not affected by the algebra error in the paper, which didn’t appear until the line R’(x) = …..   So, it looks like C is computed correctly in PureCN, at least in the place I saw within runAbsoluteCN().

Now, the only thing I’m concerned about is: (1) the seg.mean.adjusted equation you gave (above). (2) Other places in PureCN that might have copied the algebra mistake for R’(x).

As I mentioned in the bug report, I found .calcExpectedRatio() to be correct, assuming it is called with ploidy and not total.ploidy.

I updated my issue text to reflect the above.

Ted

**Comment by lima1** [comment]
Sorry about that Ted ;-)

I'll have a look and maybe add a GISTIC section in the vignette. I think the incorrect formula is never used in the package. And have a look at the C and M values in the output files. These are the purity and ploidy adjusted absolute integer copy numbers (M is the copy number of the minor "allele", M+N=C, M<=N). There shouldn't be any need for the adjusted ratio unless you need it for GISTIC and alike.,

**Comment by tedtoal** [comment]
Ok, thanks.

I’m thinking my results won’t be usable because PureCN flagged most samples as having too-low purity.  However, I’m going to plot phylogenetic trees of each tumor based on PureCN results vs. based on somatic mutations and compare.  If they are similar, I’ll have much better confidence in them.  Also, I wrote an R program to display the various PureCN solutions side by side and let user choose to override the solution for manual curation, although I haven’t run it yet to see how it goes.  I suspect it will seldom be the case that an alternate solution shows up as being obviously better.

—
Ted Toal, Postdoctoral Researcher
Carvajal-Carmona Lab
Dept. of Biochemistry and Molecular Medicine
4502 GBSF, One Shields Ave
Davis, CA  956626  (530) 263-5986
twtoal@ucdavis.edu

**Comment by lima1** [comment]
You can load the (merged) *_dnacopy.seg files in IGV and easily compare the integer copy numbers for all samples.

The low purity flag is a bit too aggressive, see https://github.com/lima1/PureCN/issues/11. It's just that below 30%, you slowly have a signal to noise ratio problem. But with high quality data, even 15% can work. But better to have a look. In an average FFPE dataset, 90% of samples are usually fine without manual curation. With small panels (<2MB) this might be closer to 80%.

**Comment by maitnnguyen** [comment]
> In some places in PureCN there are adjustments of copy ratio to account for purity and ploidy, or vice-versa, to take an adjusted value and return it to an unadjusted "observed" ratio value. Marcus Riester gave the following reference for the method used to do this:
> 
> https://www.nature.com/articles/ng.2760 (section Impurity-corrected GISTIC)
> 
> The equations in that section contain an algebra mistake. I'm concerned that PureCN might incorporate that mistake in one or more places. I have verified that it uses the correct equation in at least two places.
> 
> Details:
> 
> The above reference shows this derivation:
> 
> ```
>   R(x) = (aq(x)+2(1-a))/D
>   D = aT + 2(1-a)
>   q(x) = DR(x)/a - 2(1-a)/a
>   R'(x) = q(x)/T = R(x)/a - 2(1-a)/aT
> ```
> 
> where:
> 
> ```
>   R(x) = raw (observed) coverage ratio (PureCN's seg.mean, gene.mean, etc.)
>   R'(x) = adjusted coverage ratio (in tumor cells)
>   q(x) = integer copy number in cancer cells
>   D = average ploidy across all cells of tumor (of sample)
>   a = sample purity
>   T = tumor ploidy
> ```
> 
> However, in the last step where q(x) is substituted in q(x)/T, the algebra is wrong. The correct algebra is:
> 
> ```
> R'(x) = q(x)/T = DR(x)/aT - 2(1-a)/aT = (aT + 2(1-a))R(x)/aT - 2(1-a)/aT
>       = R(x) + 2(1-a)R(x)/aT - 2(1-a)/aT
>       = [aTR(x) + 2(1-a)R(x) - 2(1-a)]/aT
> ```
> 
> As a test, say that purity=a=0.5, tumor ploidy=T=2, and raw coverage ratio=1.5. Then we expect the adjusted coverage ratio to be 2 (tumor segment is 2X amplify (4 copies) and this becomes raw ratio of 1.5 when purity is 1/2 [0.5_4 + 0.5_2] / 2 = 1.5):
> 
> ```
> Paper:     R'(x) = 1.5/0.5 - 2(0.5)/(0.5 * 2) = 3 - 2(0.5) = 2 (correct)
> Corrected: R'(x) = [0.5*2*1.5 + 2(0.5)1.5 - 2(0.5)] / (0.5*2) = 1.5 + 1.5 - 1 = 2 (correct)
> ```
> 
> But now suppose that tumor ploidy=T=4, and we still have purity=a=0.5. Say raw coverage ratio=1.0, meaning there is no tumor amplification and the number of copies at any locus is the same as the mean number of copies, in both the 2X normal and 4X tumor tissue. We expect the adjusted coverage ratio to also be 1:
> 
> ```
> Paper:     R'(x) = 1/0.5 - 2(0.5)/(0.5 * 4) = 2 - 2(0.5)/2 = 2 - 1/2 = 1.5 (wrong)
> Corrected: R'(x) = [0.5*4*1 + 2(0.5)1 - 2(0.5)] / (0.5 * 4) = [2 + 1 - 1] / 2 = 2 / 2 = 1 (correct)
> ```
> 
> Marcus Riester indicated PureCN uses the following:
> 
> ```
>     rds <- readRDS("Sampleid.rds")
>     r <- rds$results[[1]]
>     r$seg$seg.mean.adjusted <- r$seg$seg.mean/r$purity - 2*(1-r$purity)/(r$purity*r$ploidy)
> ```
> 
> The above equation matches the incorrect one in the paper, so PureCN must be wrong wherever the above appears in the PureCN code, although I haven't found where that is (but I didn't search exhaustively).
> 
> The PureCN function .calcExpectedRatio() does the inverse operation, computing R(x) from R'(x), and it is using the correct equation.
> 
> PureCN function runAbsoluteCN() has this line:
> 
> ```
>     opt.C <- (2^(seg$seg.mean + log.ratio.offset) *  total.ploidy)/p - ((2 * (1 - p))/p)
> ```
> 
> and since tumor copy number C = tumor coverage ratio * tumor ploidy, this equation at first appears to be the paper's (incorrect) R'(x) * ploidy. However, "total.ploidy" is probably the ploidy in the sample as a whole, which corresponds to the paper's "D", and in that case, the above equation is correct. The algebra error only creeps in in the last derivation, R'(x) = ...
> 
> In summary, my concerns are (1) the above equation for seg.mean.adjusted (I don't know where in the code that is); (2) Could there be other places in PureCN where the algebra bug crept in?

Dear Tedtoal, I am using your derivation from the formula in paper that you cited, https://www.nature.com/articles/ng.2760 (section Impurity-corrected GISTIC). However, I don't know how I can cite this formula in the reference source as the formula in that paper was wrong. 
Please let me know how can I cite this derivation made by you.
Looking forward to hearing from you!
Best,
